perf/light.cc

```diff
@@ -0,0 +1,652 @@
+BENCHMARK_DEFINE_F(LightFixture, FindRandValueWithMisses)
+(benchmark::State& state)
+{
+	std::size_t lookup_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		// Re-generate dataset so it contains random values.
+		TupleHolder missing_values(state.range(0));
+		state.ResumeTiming();
+
+		for (const auto &v : missing_values.tuples) {
+			benchmark::DoNotOptimize(light_find(&ht, v.hash, v.raw));
+			lookup_count++;
+		}
+	}
+	state.SetItemsProcessed(lookup_count);
+}
+
+// Lookup random values by key.
+BENCHMARK_DEFINE_F(LightFixture, FindRandByKey)
+(benchmark::State& state)
+{
+	std::size_t lookup_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_find_key(&ht, v.hash, v.key));
+			lookup_count++;
+		}
+	}
+	state.SetItemsProcessed(lookup_count);
+}
+
+// Sequence iteration over the hash table - starting from the first value.
+// Measurements include iterator dereference.
+BENCHMARK_DEFINE_F(LightFixture, SequenceIteration)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		struct light_iterator iter;
+		light_iterator_begin(&ht, &iter);
+		Value_t *p = nullptr;
+		while ((p = light_iterator_get_and_next(&ht, &iter)) != nullptr) {
+			benchmark::DoNotOptimize(p->data);
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Fill in hash table, then erase all elements and re-fill it once again
+// (only insertion time is measured).
+BENCHMARK_DEFINE_F(LightFixture, InsertAfterErase)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		Erase(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Fill in hash table, then erase half elements and process lookups
+// (only lookup time is measured).
+BENCHMARK_DEFINE_F(LightFixture, FindAfterErase)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		// Double the data size so that after shrink it'll be of the
+		// same size as in other benches.
+		std::size_t data_size = state.range(0) << 1;
+		TupleHolder data(data_size);
+		Fill(data.tuples.begin(), data.tuples.end());
+		Erase(data.tuples.begin(), data.tuples.begin() + (data_size >> 1));
```

Comment:
Nit (please don't bother to fix now, just for your notice): better to use `std::next` for iterator arithmetics.

Reply:
Agree, fixed, thanks.